pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,200 @@
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.Deposit;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class DepositRequestManager {
+
+  private final Web3j web3j;
+  private final Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer;
+  private final DepositContract depositContract;
+  private static final Logger LOG = LogManager.getLogger();
+
+  private boolean active = false;
+  private boolean requestQueued = false;
+  private BigInteger latestCanonicalBlockNumber;
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+  private Disposable newBlockSubscription;
+
+  public DepositRequestManager(
+      Web3j web3j,
+      Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer,
+      DepositContract depositContract) {
+    this.web3j = web3j;
+    this.depositsFromBlockEventConsumer = depositsFromBlockEventConsumer;
+    this.depositContract = depositContract;
+  }
+
+  public void start() {
+    newBlockSubscription =
+        web3j
+            .blockFlowable(false)
+            .map(EthBlock::getBlock)
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(
+                this::updateLatestCanonicalBlockNumber,
+                err -> {
+                  LOG.warn("New block subscription failed, retrying.", err);
+                  start();
+                });
+  }
+
+  public void stop() {
+    newBlockSubscription.dispose();
+  }
+
+  private synchronized void updateLatestCanonicalBlockNumber(
+      BigInteger latestCanonicalBlockNumber) {
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    if (active) {
+      requestQueued = true;
+    } else {
+      active = true;
+      getLatestDeposits();
+    }
+  }
+
+  private void getLatestDeposits() {
+    requestQueued = false;
+
+    DefaultBlockParameter fromBlock =
+        DefaultBlockParameter.valueOf(latestSuccessfullyQueriedBlockNumber);
+    final BigInteger latestCanonicalBlockNumberAtRequestStart = this.latestCanonicalBlockNumber;
+    DefaultBlockParameter toBlock =
+        DefaultBlockParameter.valueOf(latestCanonicalBlockNumberAtRequestStart);
+
+    LOG.trace(
+        "Attempting to fetch deposit events for block numbers in the range ({}, {})",
+        latestSuccessfullyQueriedBlockNumber,
+        latestCanonicalBlockNumberAtRequestStart);
+
+    fetchAndPublishDepositEventsInBlockRange(fromBlock, toBlock)
+        .finish(
+            () -> {
+              this.latestSuccessfullyQueriedBlockNumber = latestCanonicalBlockNumberAtRequestStart;
+              synchronized (DepositRequestManager.this) {
+                active = false;
+                if (requestQueued) {
+                  getLatestDeposits();
+                }
+              }
```

Comment:
I'd be tempted to extract this as an `onDepositRequestSuccessful` method.

Reply:
Done.